WordPress/gutenberg
# Experimental Link creation interface

This PR is a working prototype for a potential new UI for adding/editing links. 

~It is a WIP and is **not ready for review.**~ 

Closes https://github.com/WordPress/gutenberg/issues/17557

## How has this been tested?
* Automated tests
* Manually testing in the Playground

## Screenshots 

![Screen Capture on 2019-10-16 at 11-34-37](https://user-images.githubusercontent.com/444434/66911738-07ef3580-f009-11e9-9b91-cdddd777182f.gif)

## Todo
- [x] Fix `URLInput` tests
- [ ] Document changes to `URLInput`.
- [ ] Increase test coverage for `LinkControl`.
- [x] Handle `mailto` and `tel` protocol plus internal anchors.
- [x] Fix console error when hitting `ENTER` key when the text input has a value and is focused. 
- [x] Expose Posts, Pages and Categories in the default search results - as per [this convo](https://github.com/WordPress/gutenberg/issues/17557#issuecomment-540571946).~
- [x] ~Add publish date information to data fetch calls for search - currently, we only get bare-bones info about the found posts (see also above as may be related). Could we [add publish date to the REST API in Core](https://github.com/WordPress/WordPress/blob/c6d3987b0831412f18ace8ea0f5c5423107a0d55/wp-includes/rest-api/search/class-wp-rest-post-search-handler.php#L104)?~ Decided this was not required.
- [x] Add `aria-hidden` to the appropriate elements for entity-based results to avoid URL being read out to screen-readers.
- [x] Fix scrolling shadows on search suggestions. Looks odd in Safari. Maybe try [this technique](
https://css-scroll-shadows.now.sh/?bgColor=bad6fc&shadowColor=222222&pxSize=15).
- [x] Implement [additional toggle draw for additional "additional settings"]
- [ ] Avoid inline component pattern [for perf reasons]
- [x] Improve URL detection mechanic to [recognise URLs that don't start with `http` or `https`](https://github.com/WordPress/gutenberg/pull/17846#issuecomment-543244810).
(https://github.com/WordPress/gutenberg/pull/17846#discussion_r335371187).
- [x] The loading spinner makes the experience feel jumpy, is there a position for this that doesn't cause items below to move up and down as it toggles visibility?
- [x] [Tweak `line-height` up a little](https://github.com/WordPress/gutenberg/pull/17846#issuecomment-542853583) on `url` within search suggestion items.
- [x] [Restore visual indication of focus state](https://github.com/WordPress/gutenberg/pull/17846) when moving between items using keyboard.
- [x] Add into Nav Block for in situ testing.
- [x] [Solve screen reader issue where the entire suggestion content is read out](https://github.com/WordPress/gutenberg/pull/17846#issuecomment-543014076) in a single sentence.
- [x] Truncate URLs for _entity_ search suggestions and have them overflow with ellipsis rather than wrap. 
- [x] Fix broken `URLInput` tests and check existing implementations across UI to ensure we haven't broken anything.
- [x] Document changes to `URLInput`.
- [x] Capitalise `URL` in the search results.
- [x] Implement ability to select an option.
- [x] Ability to remove a selected option.
- [x] Ability to edit a selected option. 
- [x] Fix overflow issue when extra long url is entered. Because word does not break we need to handle this carefully.
- [x] Remove "trigger" button UI from the code - the current use cases indicated mean that this component should really only expose the link creation interface.
- [x] Implement "overflow" scroll on search results when many results are rendered.
- [x] Remove "insert" button on search results and reimplement `<button>`s for each result. 
- [x] Test whether categories are returned by default in the search mechanic (doubtful). Determine whether this is actually a requirement or whether we should simply ensure we are exposing an API to allow utilisation of a custom post fetch mechanic.
- [x] Implement URL detection / parsing mechanic to distinguish URL entry from searches.
- [x] Implement URL "object" creation.
- [x] Implement "additional settings" UI as shown in designs.
(https://github.com/WordPress/gutenberg/issues/17557#issuecomment-540332745).
- [x] Fix bug where resetting input value doesn't clear the suggestions.
- [x] ~[Determine whether it's feasible](https://github.com/WordPress/gutenberg/issues/17557#issuecomment-540659203) (for an MVP) to have URL search do a lookup of the entered site's `<title>` and favicon.~ [We agreed this wasn't a hard requirement](https://github.com/WordPress/gutenberg/issues/17557#issuecomment-541125677) for an MVP.
- [x] Implement search term highlighting within search matches (eg: search is for "Foo" then the search result for `Foobar Page` should be formatted as "**Foo**bar Page")

## Testing Instructions

You can test this by 

* checking out this PR.
* running `npm run dev` - leave this running and switch to a new terminal tab
* [in a new terminal tab] running `npm run playground:dev` from the Gutenberg repo root.
* you should see the new UI by going to `http://localhost:1234`





## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
